### ISSUE TITLE
MultiColor

### ISSUE THREAD

**Issue by purple-movies** [mozilla-iot/sengled-adapter#6]
Hi there, 

I'm new to writing WebThings adapters & I was wondering if it's possible to add support for Sengled multi-color bulbs ?

**Comment by mrstegeman** [comment]
It's certainly possible, but...
1. I don't have a color bulb to test with.
2. I had to reverse engineer the Android app to get this add-on to work.
3. I had to watch the traffic to Sengled's cloud to figure out the protocol.
4. I can't watch the traffic without a color bulb... See point 1.

I'd love to help, but I'm unable to right now.

**Comment by purple-movies** [comment]
Wow - thanks for the quick response. 

I might be able to figure it out, but there's a bit of a learning curve on how the adapters work, & I haven't worked with Python a lot. 

Are there any good adapter tutorials you know about ? The ones I could find seemed a bit vague. 

Also if you could point me to where it's interacting with the Sengled cloud it might help. I think I found it, just unfamiliar with the code.

**Comment by mrstegeman** [comment]
There are lots of good resources here: https://github.com/WebThingsIO/wiki/wiki#general-1

In the code:
* Client: https://github.com/WebThingsIO/sengled-adapter/blob/master/pkg/client.py
* Device (where you'll add the color property and such): https://github.com/WebThingsIO/sengled-adapter/blob/master/pkg/sengled_device.py
* Property (where you'll handle changes to the property): https://github.com/WebThingsIO/sengled-adapter/blob/master/pkg/sengled_property.py#L28-L33

**Comment by purple-movies** [comment]
Thanks - I'll see if I can get somewhere with that tonight.

**Comment by mrstegeman** [comment]
Definitely let me know if you need more help!

**Comment by purple-movies** [comment]
TBH - I'm still finding it a bit unclear how to test/debug any changes I make with WebThings. 

Is this a process of copying the files with changes over to a storage location on the Pi running WebThings & restarting it ? If so, what's the best way to debug ?

Otherwise, is there a better way to develop adapters ?

**Comment by mrstegeman** [comment]
You can develop on the Pi itself if you want. If you want to develop on a separate machine, then yes, you'll need to make your changes and then copy the files over. After doing so, you can just restart the individual add-on from the UI by navigating to _Settings -> Add-ons_ and Disabling, then Enabling the add-on.

**Comment by DaveGut** [comment]
I am developing a MQTT interface to the color bulb.  The following works for COLOR.  The following works

        data = {
            'dn': self.uuid,
            'type': 'color',
            'value': str(red:green:blue),
            'time': int(time.time() * 1000),
        }
red, green, and blue are all ranged from 0 -255

PS.  I do not use time (tested both ways without issues).  Example of the data after converted to JSON is below:

**{"dn":"80:A0:36:E1:75:AA","type":"color","value":"196:165:55"}**

The following works for COLORTEMPERATURE:

        data = {
            'dn': self.uuid,
            'type': 'colorTemperature',
            'value': str(percent of range),
            'time': int(time.time() * 1000),
        }
Standard color temp range in Kelvin is 2000 to 4500.  Therefore,

- precent of range = ( (colorTemp - 2000) / 45 ).toInteger()

Example: set color temp to 5000 (66% of range) after converting to JSON
**{"dn":"80:A0:36:E1:75:67","type":"colorTemperature","value":"66"}**

**Comment by purple-movies** [comment]
@DaveGut Is that something you're adding to the adapter ?

**Comment by DaveGut** [comment]
No.  I use Hubitat with a MQTT implementation.  That is how I tested the commands for the wifi color bulb.
